### ISSUE TITLE
Uploading gif causes freezing/ slow-down/ crashing

### ISSUE THREAD

**Issue by jwzimmer** [LLK/scratch-flash#1238]
A user reported via freshdesk no. 61590:
> I write to you today to tell of a peculiar glitch: Every time I try to upload a .GIF file, shockwave flash crashes, or, on my significantly older Mac, I get the black-screen-of-death and have to unplug the computer! 
> The first computer I tried it on was on google chrome, recent adobe flash, and Windows 7. The OTHER computer though is a 9 year old IMac, and I wouldn't worry about it because it often gets messed-up by a lot of my more advanced projects. 

They included the gifs in the freshdesk ticket, so the files are there. I uploaded the files to my project 132100241. I was able to upload the gifs, but I did get a slow-down and message about that:
<img width="1251" alt="screen shot 2016-11-21 at 3 50 02 pm" src="https://cloud.githubusercontent.com/assets/1320456/20500504/a3fcef60-b003-11e6-8393-fd97e363bf7f.png">

@cwillisf and I discussed this, and he thought it would be worth looking into what's causing the slow-down, even if it might be that we can't do anything about it, or that the change would be more work than is worthwhile in 2.0 at this point.

**Comment by cwillisf** [comment]
I took a quick look at this just now, and there are two main hotspots:
- We spend about 2/3 of the time decoding the GIF. It seems that our `GIFDecoder` class is somewhat slow... maybe we can replace it with a third-party library or browser / Node function in Scratch 3.0
- We spend about 1/3 of the time creating `ScratchCostume` objects. Most of that time is spent re-compressing the image using PNG rather than GIF compression. This time is spent inside the `encode` method of the third-party PNG library that we're using. This could (probably should) be delayed until save time, but there's no obvious way to make it take less time.

Bottom line: it doesn't look there's any low-hanging fruit here. I think improvements would mean either replacing these methods with other, higher-performance third-party libraries or writing new, performance-minded GIF decode and PNG encode methods in AS3. I propose that we do neither of these things right now, and instead make sure we do better than this in Scratch 3.0
